do/firewalls.go

```diff
@@ -0,0 +1,235 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package do
+
+import (
+	"fmt"
+	"github.com/digitalocean/godo"
+	"github.com/digitalocean/godo/context"
+	"strconv"
+	"strings"
+)
+
+type firewallRequest struct {
+	firewallID string
+	rules      *godo.FirewallRulesRequest
+}
+
+func checkIfPortInRange(port int, fwPortRange string) bool {
+
+	fwPortList := strings.Split(fwPortRange, "-")
+	if fwPortList[0] == "all" {
+		return true
+	}
+	if len(fwPortList) == 1 {
+		fwPort, _ := strconv.Atoi(fwPortList[0])
+		if port == fwPort {
+			return true
+		}
+	} else {
+		fwLowerVal, _ := strconv.Atoi(fwPortList[0])
+		fwUpperVal, _ := strconv.Atoi(fwPortList[1])
+		if port >= fwLowerVal && port <= fwUpperVal {
+			return true
+		}
+	}
+
+	return false
+
+}
+
+func checkIfPortRequiresDeletion(port int, fwPortRange string) bool {
+
+	fwPortList := strings.Split(fwPortRange, "-")
+	if fwPortList[0] == "all" {
+		return false
+	}
+	if len(fwPortList) == 1 {
+		fwPort, _ := strconv.Atoi(fwPortList[0])
+		if port == fwPort {
+			return true
+		}
+	}
+
+	return false
+
+}
+
+func createRuleRequest(sourcelb string, targetPort int) *godo.FirewallRulesRequest {
+	rr := &godo.FirewallRulesRequest{
+		InboundRules: []godo.InboundRule{
+			{
+				Protocol:  "tcp",
+				PortRange: strconv.Itoa(targetPort),
+				Sources: &godo.Sources{
+					LoadBalancerUIDs: []string{sourcelb},
+				},
+			},
+		},
+	}
+	return rr
+}
+
+// Check if a firewall rule already exists for a load balancer
+// forwarding rule ( returns true if covered by ALL, a port range or a single port)
+func lbfwRuleExists(lbID string, lbRule godo.ForwardingRule, firewalls []godo.Firewall) bool {
+	ruleExists := false
+	//check each firewall to see if rule already exists
+	for _, firewall := range firewalls {
+		for _, fwInboundRule := range firewall.InboundRules {
+			for _, fwSourcelbs := range fwInboundRule.Sources.LoadBalancerUIDs {
+				if fwSourcelbs == lbID {
+					res := checkIfPortInRange(lbRule.TargetPort, fwInboundRule.PortRange)
+					if res && fwInboundRule.Protocol == "tcp" {
+						ruleExists = true
+					}
+				}
+			}
+		}
+	}
+
+	return ruleExists
+
+}
+
+func requestExists(requests []firewallRequest, firewallID string, lbID string, targetPort int) bool {
+
+	for _, request := range requests {
+		if firewallID == request.firewallID &&
+			lbID == request.rules.InboundRules[0].Sources.LoadBalancerUIDs[0] &&
+			strconv.Itoa(targetPort) == request.rules.InboundRules[0].PortRange {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Check if a single port rule exists for a load balancer
+// forwarding rule ( will not remove ranges defined by ports or ALL )
+func lbfwRuleRequiresDeletion(lbID string, lbRule godo.ForwardingRule, firewalls []godo.Firewall) bool {
```

Comment:
This function is pretty much the same as lbfwRuleExists, only difference is it makes a call to checkIfPortRequiresDeletion, rather than checkIfPortInRange. These are slightly different as we only delete a port if it's not included in a range of ports all defined with "ALL". Whereas we only add a port if it's not contained within a list of ranges. 

These two functions (lbfwRuleExists, lbfwRuleRequiresDeletion) could be replaced by one function, and the checking function required could be passed in as a parameter. Although I'm not sure if that is good practice.